File: frontend/src/types/mtp.ts
// MTP (Module Type Package) type definitions
// Mirrors backend/shared/src/mtp.rs — keep in sync

// ─── PEA Configuration ──────────────────────────────────────────────────────

export interface PeaConfig {
  id: string
  name: string
  version: string
  description: string
  writer: WriterInfo
  services: ServiceConfig[]
  active_elements: ActiveElement[]
  opcua_config: OpcUaConfig
  created_at: string
  updated_at: string
}

export interface WriterInfo {
  name: string
  version: string
  vendor: string
}

export interface OpcUaConfig {
  endpoint: string
  namespace_uri: string
  security_policy: string
}

// ─── Service ─────────────────────────────────────────────────────────────────

export interface ServiceConfig {
  tag: string
  name: string
  description: string
  config_parameters: ServiceParameter[]
  procedures: ProcedureConfig[]
}

// ─── Procedure ───────────────────────────────────────────────────────────────

export interface ProcedureConfig {
  id: number
  name: string
  is_self_completing: boolean
  is_default: boolean
  parameters: ServiceParameter[]
  process_value_outs: IndicatorElement[]
  report_values: IndicatorElement[]
}

// ─── Parameter Types (MTP Operation Elements) ────────────────────────────────

export type ServiceParameter =
  | ({ type: 'Analog' } & AnalogParameter)
  | ({ type: 'Binary' } & BinaryParameter)
  | ({ type: 'DInt' } & DIntParameter)
  | ({ type: 'StringParam' } & StringParameter)

export interface AnalogParameter {
  tag: string
  name: string
  unit: string
  v_scl_min: number
  v_scl_max: number
  v_min: number
  v_max: number
  v_default: number
  tag_mapping: TagMapping | null
}

export interface BinaryParameter {
  tag: string
  name: string
  v_state0: string
  v_state1: string
  v_default: boolean
  tag_mapping: TagMapping | null
}

export interface DIntParameter {
  tag: string
  name: string
  unit: string
  v_scl_min: number
  v_scl_max: number
  v_min: number
  v_max: number
  v_default: number
  tag_mapping: TagMapping | null
}

export interface StringParameter {
  tag: string
  name: string
  v_default: string
  tag_mapping: TagMapping | null
}

// ─── Indicator Elements (MTP Views) ──────────────────────────────────────────

export type IndicatorElement =
  | ({ type: 'AnaView' } & AnaViewConfig)
  | ({ type: 'BinView' } & BinViewConfig)
  | ({ type: 'BinStringView' } & BinStringViewConfig)
  | ({ type: 'DIntView' } & DIntViewConfig)
  | ({ type: 'DIntStringView' } & DIntStringViewConfig)
  | ({ type: 'StringView' } & StringViewConfig)

export interface AnaViewConfig {
  tag: string
  name: string
  unit: string
  v_scl_min: number
  v_scl_max: number
  tag_mapping: TagMapping | null
}

export interface BinViewConfig {
  tag: string
  name: string
  v_state0: string
  v_state1: string
  tag_mapping: TagMapping | null
}

export interface BinStringViewConfig {
  tag: string
  name: string
  v_state0: string
  v_state1: string
  tag_mapping: TagMapping | null
}

export interface DIntViewConfig {
  tag: string
  name: string
  unit: string
  v_scl_min: number
  v_scl_max: number
  tag_mapping: TagMapping | null
}

export interface DIntStringViewConfig {
  tag: string
  name: string
  v_scl_min: number
  v_scl_max: number
  tag_mapping: TagMapping | null
}

export interface StringViewConfig {
  tag: string
  name: string
  tag_mapping: TagMapping | null
}

// ─── Active Elements ─────────────────────────────────────────────────────────

export type ActiveElement =
  | ({ element_type: 'BinVlv' } & BinVlvConfig)
  | ({ element_type: 'BinMon' } & BinMonConfig)
  | ({ element_type: 'AnaVlv' } & AnaVlvConfig)
  | ({ element_type: 'BinDrv' } & BinDrvConfig)
  | ({ element_type: 'AnaDrv' } & AnaDrvConfig)
  | ({ element_type: 'DIntDrv' } & DIntDrvConfig)
  | ({ element_type: 'DIntMon' } & DIntMonConfig)
  | ({ element_type: 'PIDCtrl' } & PIDCtrlConfig)

export interface BinVlvConfig {
  tag: string
  name: string
  safe_pos: boolean
  open_fbk_tag: TagMapping | null
  close_fbk_tag: TagMapping | null
  open_cmd_tag: TagMapping | null
  close_cmd_tag: TagMapping | null
}

export interface BinMonConfig {
  tag: string
  name: string
  fbk_tag: TagMapping | null
}

export interface AnaVlvConfig {
  tag: string
  name: string
  safe_pos: number
  pos_min: number
  pos_max: number
  pos_unit: string
  pos_fbk_tag: TagMapping | null
  pos_sp_tag: TagMapping | null
}

export interface BinDrvConfig {
  tag: string
  name: string
  safe_pos: boolean
  fwd_fbk_tag: TagMapping | null
  rev_fbk_tag: TagMapping | null
  fwd_cmd_tag: TagMapping | null
  rev_cmd_tag: TagMapping | null
  stop_cmd_tag: TagMapping | null
}

export interface AnaDrvConfig {
  tag: string
  name: string
  safe_pos: number
  rpm_min: number
  rpm_max: number
  rpm_unit: string
  rpm_fbk_tag: TagMapping | null
  rpm_sp_tag: TagMapping | null
  fwd_cmd_tag: TagMapping | null
  rev_cmd_tag: TagMapping | null
  stop_cmd_tag: TagMapping | null
}

export interface DIntDrvConfig {
  tag: string
  name: string
  safe_pos: number
  rpm_min: number
  rpm_max: number
  rpm_unit: string
  rpm_fbk_tag: TagMapping | null
  rpm_sp_tag: TagMapping | null
  fwd_cmd_tag: TagMapping | null
  rev_cmd_tag: TagMapping | null
  stop_cmd_tag: TagMapping | null
}

export interface DIntMonConfig {
  tag: string
  name: string
  unit: string
  v_scl_min: number
  v_scl_max: number
  fbk_tag: TagMapping | null
}

export interface PIDCtrlConfig {
  tag: string
  name: string
  kp: number
  ki: number
  kd: number
  pv_unit: string
  pv_scl_min: number
  pv_scl_max: number
  sp_scl_min: number
  sp_scl_max: number
  mv_scl_min: number
  mv_scl_max: number
  pv_tag: TagMapping | null
  sp_tag: TagMapping | null
  mv_tag: TagMapping | null
}

// ─── Tag Mapping ─────────────────────────────────────────────────────────────

export interface TagMapping {
  protocol: ProtocolType
  address: string
}

export type ProtocolType = 'OpcUa' | 'Modbus' | 'Zenoh'

// ─── PackML Service State Machine ────────────────────────────────────────────

export enum ServiceState {
  Idle = 'Idle',
  Starting = 'Starting',
  Execute = 'Execute',
  Completing = 'Completing',
  Completed = 'Completed',
  Pausing = 'Pausing',
  Paused = 'Paused',
  Resuming = 'Resuming',
  Holding = 'Holding',
  Held = 'Held',
  Unholding = 'Unholding',
  Stopping = 'Stopping',
  Stopped = 'Stopped',
  Aborting = 'Aborting',
  Aborted = 'Aborted',
  Resetting = 'Resetting',
}

export const SERVICE_STATE_CODES: Record<ServiceState, number> = {
  [ServiceState.Idle]: 16,
  [ServiceState.Starting]: 8,
  [ServiceState.Execute]: 64,
  [ServiceState.Completing]: 65536,
  [ServiceState.Completed]: 131072,
  [ServiceState.Pausing]: 8192,
  [ServiceState.Paused]: 32,
  [ServiceState.Resuming]: 16384,
  [ServiceState.Holding]: 1024,
  [ServiceState.Held]: 2048,
  [ServiceState.Unholding]: 4096,
  [ServiceState.Stopping]: 128,
  [ServiceState.Stopped]: 4,
  [ServiceState.Aborting]: 256,
  [ServiceState.Aborted]: 512,
  [ServiceState.Resetting]: 32768,
}

export enum ServiceCommand {
  Reset = 'Reset',
  Start = 'Start',
  Stop = 'Stop',
  Hold = 'Hold',
  Unhold = 'Unhold',
  Pause = 'Pause',
  Resume = 'Resume',
  Abort = 'Abort',
  Restart = 'Restart',
  Complete = 'Complete',
}

export const ALLOWED_COMMANDS: Record<ServiceState, ServiceCommand[]> = {
  [ServiceState.Idle]: [ServiceCommand.Start, ServiceCommand.Abort],
  [ServiceState.Execute]: [
    ServiceCommand.Complete, ServiceCommand.Hold,
    ServiceCommand.Pause, ServiceCommand.Stop, ServiceCommand.Abort,
  ],
  [ServiceState.Completed]: [ServiceCommand.Reset, ServiceCommand.Stop, ServiceCommand.Abort],
  [ServiceState.Paused]: [ServiceCommand.Resume, ServiceCommand.Stop, ServiceCommand.Abort],
  [ServiceState.Held]: [ServiceCommand.Unhold, ServiceCommand.Stop, ServiceCommand.Abort],
  [ServiceState.Stopped]: [ServiceCommand.Reset, ServiceCommand.Abort],
  [ServiceState.Aborted]: [ServiceCommand.Reset],
  // Transient states: no commands
  [ServiceState.Starting]: [],
  [ServiceState.Completing]: [],
  [ServiceState.Pausing]: [],
  [ServiceState.Resuming]: [],
  [ServiceState.Holding]: [],
  [ServiceState.Unholding]: [],
  [ServiceState.Stopping]: [],
  [ServiceState.Aborting]: [],
  [ServiceState.Resetting]: [],
}

export type OperationMode = 'Offline' | 'Operator' | 'Automatic'
export type SourceMode = 'Internal' | 'External'

// ─── PEA Instance Runtime Status ─────────────────────────────────────────────

export interface PeaInstanceStatus {
  pea_id: string
  deployed: boolean
  running: boolean
  services: ServiceRuntimeState[]
  opcua_endpoint: string | null
  last_updated: string
}

export interface ServiceRuntimeState {
  tag: string
  state: ServiceState
  current_procedure_id: number | null
  operation_mode: OperationMode
  source_mode: SourceMode
}

// ─── State Color Helpers ─────────────────────────────────────────────────────

export function getStateColor(state: ServiceState): 'success' | 'warning' | 'error' | 'info' | 'default' {
  switch (state) {
    case ServiceState.Execute:
    case ServiceState.Completed:
      return 'success'
    case ServiceState.Idle:
    case ServiceState.Held:
    case ServiceState.Paused:
      return 'warning'
    case ServiceState.Aborted:
    case ServiceState.Stopped:
    case ServiceState.Aborting:
    case ServiceState.Stopping:
      return 'error'
    case ServiceState.Starting:
    case ServiceState.Resuming:
    case ServiceState.Unholding:
    case ServiceState.Resetting:
      return 'info'
    default:
      return 'default'
  }
}

// ─── Zenoh Topic Helpers ─────────────────────────────────────────────────────

export const ZENOH_TOPICS = {
  peaAnnounce: (peaId: string) => `fendtastic/pea/${peaId}/announce`,
  peaStatus: (peaId: string) => `fendtastic/pea/${peaId}/status`,
  peaServiceState: (peaId: string, svcTag: string) =>
    `fendtastic/pea/${peaId}/services/${svcTag}/state`,
  peaServiceCommand: (peaId: string, svcTag: string) =>
    `fendtastic/pea/${peaId}/services/${svcTag}/command`,
  peaData: (peaId: string, dataTag: string) =>
    `fendtastic/pea/${peaId}/data/${dataTag}`,
  peaConfig: (peaId: string) => `fendtastic/pea/${peaId}/config`,

  peaDiscoveryWildcard: 'fendtastic/pea/+/announce',
  peaStatusWildcard: 'fendtastic/pea/+/status',
  polRecipesCommand: 'fendtastic/pol/recipes/command',
  polRecipesStatus: 'fendtastic/pol/recipes/status',
  statusEvaIcs: 'fendtastic/status/eva-ics',
} as const

// ─── Factory Helpers ─────────────────────────────────────────────────────────

export function createEmptyPeaConfig(): PeaConfig {
  const now = new Date().toISOString()
  return {
    id: '',
    name: '',
    version: '1.0.0',
    description: '',
    writer: { name: '', version: '1.0.0', vendor: '' },
    services: [],
    active_elements: [],
    opcua_config: {
      endpoint: 'opc.tcp://0.0.0.0:4840',
      namespace_uri: '',
      security_policy: 'None',
    },
    created_at: now,
    updated_at: now,
  }
}

export function createEmptyService(): ServiceConfig {
  return {
    tag: '',
    name: '',
    description: '',
    config_parameters: [],
    procedures: [{
      id: 1,
      name: 'Default',
      is_self_completing: false,
      is_default: true,
      parameters: [],
      process_value_outs: [],
      report_values: [],
    }],
  }
}

export function createEmptyProcedure(id: number): ProcedureConfig {
  return {
    id,
    name: '',
    is_self_completing: false,
    is_default: false,
    parameters: [],
    process_value_outs: [],
    report_values: [],
  }
}

File: backend/shared/src/mtp.rs
use serde::{Deserialize, Serialize};

// ─── PEA Information Label ───────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeaConfig {
    pub id: String,
    pub name: String,
    pub version: String,
    pub description: String,
    pub writer: WriterInfo,
    pub services: Vec<ServiceConfig>,
    pub active_elements: Vec<ActiveElement>,
    pub opcua_config: OpcUaConfig,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WriterInfo {
    pub name: String,
    pub version: String,
    pub vendor: String,
}

// ─── OPC UA Configuration ────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpcUaConfig {
    pub endpoint: String,
    pub namespace_uri: String,
    pub security_policy: String,
}

// ─── Service ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceConfig {
    pub tag: String,
    pub name: String,
    pub description: String,
    pub config_parameters: Vec<ServiceParameter>,
    pub procedures: Vec<ProcedureConfig>,
}

// ─── Procedure ───────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcedureConfig {
    pub id: u32,
    pub name: String,
    pub is_self_completing: bool,
    pub is_default: bool,
    pub parameters: Vec<ServiceParameter>,
    pub process_value_outs: Vec<IndicatorElement>,
    pub report_values: Vec<IndicatorElement>,
}

// ─── Parameter Types (MTP Operation Elements) ────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServiceParameter {
    Analog(AnalogParameter),
    Binary(BinaryParameter),
    DInt(DIntParameter),
    StringParam(StringParameter),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnalogParameter {
    pub tag: String,
    pub name: String,
    pub unit: String,
    pub v_scl_min: f64,
    pub v_scl_max: f64,
    pub v_min: f64,
    pub v_max: f64,
    pub v_default: f64,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinaryParameter {
    pub tag: String,
    pub name: String,
    pub v_state0: String,
    pub v_state1: String,
    pub v_default: bool,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DIntParameter {
    pub tag: String,
    pub name: String,
    pub unit: String,
    pub v_scl_min: i64,
    pub v_scl_max: i64,
    pub v_min: i64,
    pub v_max: i64,
    pub v_default: i64,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StringParameter {
    pub tag: String,
    pub name: String,
    pub v_default: String,
    pub tag_mapping: Option<TagMapping>,
}

// ─── Indicator Elements (MTP Views) ──────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum IndicatorElement {
    AnaView(AnaViewConfig),
    BinView(BinViewConfig),
    BinStringView(BinStringViewConfig),
    DIntView(DIntViewConfig),
    DIntStringView(DIntStringViewConfig),
    StringView(StringViewConfig),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnaViewConfig {
    pub tag: String,
    pub name: String,
    pub unit: String,
    pub v_scl_min: f64,
    pub v_scl_max: f64,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinViewConfig {
    pub tag: String,
    pub name: String,
    pub v_state0: String,
    pub v_state1: String,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinStringViewConfig {
    pub tag: String,
    pub name: String,
    pub v_state0: String,
    pub v_state1: String,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DIntViewConfig {
    pub tag: String,
    pub name: String,
    pub unit: String,
    pub v_scl_min: i64,
    pub v_scl_max: i64,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DIntStringViewConfig {
    pub tag: String,
    pub name: String,
    pub v_scl_min: i64,
    pub v_scl_max: i64,
    pub tag_mapping: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StringViewConfig {
    pub tag: String,
    pub name: String,
    pub tag_mapping: Option<TagMapping>,
}

// ─── Active Elements ─────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "element_type")]
pub enum ActiveElement {
    BinVlv(BinVlvConfig),
    BinMon(BinMonConfig),
    AnaVlv(AnaVlvConfig),
    BinDrv(BinDrvConfig),
    AnaDrv(AnaDrvConfig),
    DIntDrv(DIntDrvConfig),
    DIntMon(DIntMonConfig),
    PIDCtrl(PIDCtrlConfig),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinVlvConfig {
    pub tag: String,
    pub name: String,
    pub safe_pos: bool,
    pub open_fbk_tag: Option<TagMapping>,
    pub close_fbk_tag: Option<TagMapping>,
    pub open_cmd_tag: Option<TagMapping>,
    pub close_cmd_tag: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinMonConfig {
    pub tag: String,
    pub name: String,
    pub fbk_tag: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnaVlvConfig {
    pub tag: String,
    pub name: String,
    pub safe_pos: f64,
    pub pos_min: f64,
    pub pos_max: f64,
    pub pos_unit: String,
    pub pos_fbk_tag: Option<TagMapping>,
    pub pos_sp_tag: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BinDrvConfig {
    pub tag: String,
    pub name: String,
    pub safe_pos: bool,
    pub fwd_fbk_tag: Option<TagMapping>,
    pub rev_fbk_tag: Option<TagMapping>,
    pub fwd_cmd_tag: Option<TagMapping>,
    pub rev_cmd_tag: Option<TagMapping>,
    pub stop_cmd_tag: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnaDrvConfig {
    pub tag: String,
    pub name: String,
    pub safe_pos: f64,
    pub rpm_min: f64,
    pub rpm_max: f64,
    pub rpm_unit: String,
    pub rpm_fbk_tag: Option<TagMapping>,
    pub rpm_sp_tag: Option<TagMapping>,
    pub fwd_cmd_tag: Option<TagMapping>,
    pub rev_cmd_tag: Option<TagMapping>,
    pub stop_cmd_tag: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DIntDrvConfig {
    pub tag: String,
    pub name: String,
    pub safe_pos: i64,
    pub rpm_min: i64,
    pub rpm_max: i64,
    pub rpm_unit: String,
    pub rpm_fbk_tag: Option<TagMapping>,
    pub rpm_sp_tag: Option<TagMapping>,
    pub fwd_cmd_tag: Option<TagMapping>,
    pub rev_cmd_tag: Option<TagMapping>,
    pub stop_cmd_tag: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DIntMonConfig {
    pub tag: String,
    pub name: String,
    pub unit: String,
    pub v_scl_min: i64,
    pub v_scl_max: i64,
    pub fbk_tag: Option<TagMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PIDCtrlConfig {
    pub tag: String,
    pub name: String,
    pub kp: f64,
    pub ki: f64,
    pub kd: f64,
    pub pv_unit: String,
    pub pv_scl_min: f64,
    pub pv_scl_max: f64,
    pub sp_scl_min: f64,
    pub sp_scl_max: f64,
    pub mv_scl_min: f64,
    pub mv_scl_max: f64,
    pub pv_tag: Option<TagMapping>,
    pub sp_tag: Option<TagMapping>,
    pub mv_tag: Option<TagMapping>,
}

// ─── Tag Mapping ─────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TagMapping {
    pub protocol: ProtocolType,
    pub address: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ProtocolType {
    OpcUa,
    Modbus,
    Zenoh,
}

// ─── PackML Service State Machine ────────────────────────────────────────────

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum ServiceState {
    Idle,
    Starting,
    Execute,
    Completing,
    Completed,
    Pausing,
    Paused,
    Resuming,
    Holding,
    Held,
    Unholding,
    Stopping,
    Stopped,
    Aborting,
    Aborted,
    Resetting,
}

impl ServiceState {
    pub fn code(&self) -> u32 {
        match self {
            Self::Idle => 16,
            Self::Starting => 8,
            Self::Execute => 64,
            Self::Completing => 65536,
            Self::Completed => 131072,
            Self::Pausing => 8192,
            Self::Paused => 32,
            Self::Resuming => 16384,
            Self::Holding => 1024,
            Self::Held => 2048,
            Self::Unholding => 4096,
            Self::Stopping => 128,
            Self::Stopped => 4,
            Self::Aborting => 256,
            Self::Aborted => 512,
            Self::Resetting => 32768,
        }
    }

    pub fn from_code(code: u32) -> Option<Self> {
        match code {
            16 => Some(Self::Idle),
            8 => Some(Self::Starting),
            64 => Some(Self::Execute),
            65536 => Some(Self::Completing),
            131072 => Some(Self::Completed),
            8192 => Some(Self::Pausing),
            32 => Some(Self::Paused),
            16384 => Some(Self::Resuming),
            1024 => Some(Self::Holding),
            2048 => Some(Self::Held),
            4096 => Some(Self::Unholding),
            128 => Some(Self::Stopping),
            4 => Some(Self::Stopped),
            256 => Some(Self::Aborting),
            512 => Some(Self::Aborted),
            32768 => Some(Self::Resetting),
            _ => None,
        }
    }

    pub fn is_stable(&self) -> bool {
        matches!(
            self,
            Self::Idle
                | Self::Execute
                | Self::Completed
                | Self::Paused
                | Self::Held
                | Self::Stopped
                | Self::Aborted
        )
    }

    pub fn allowed_commands(&self) -> Vec<ServiceCommand> {
        match self {
            Self::Idle => vec![ServiceCommand::Start, ServiceCommand::Abort],
            Self::Execute => vec![
                ServiceCommand::Complete,
                ServiceCommand::Hold,
                ServiceCommand::Pause,
                ServiceCommand::Stop,
                ServiceCommand::Abort,
            ],
            Self::Completed => vec![
                ServiceCommand::Reset,
                ServiceCommand::Stop,
                ServiceCommand::Abort,
            ],
            Self::Paused => vec![
                ServiceCommand::Resume,
                ServiceCommand::Stop,
                ServiceCommand::Abort,
            ],
            Self::Held => vec![
                ServiceCommand::Unhold,
                ServiceCommand::Stop,
                ServiceCommand::Abort,
            ],
            Self::Stopped => vec![ServiceCommand::Reset, ServiceCommand::Abort],
            Self::Aborted => vec![ServiceCommand::Reset],
            // Transient states: no commands allowed
            _ => vec![],
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum ServiceCommand {
    Reset,
    Start,
    Stop,
    Hold,
    Unhold,
    Pause,
    Resume,
    Abort,
    Restart,
    Complete,
}

impl ServiceCommand {
    pub fn code(&self) -> u32 {
        match self {
            Self::Reset => 2,
            Self::Start => 4,
            Self::Stop => 8,
            Self::Hold => 16,
            Self::Unhold => 32,
            Self::Pause => 64,
            Self::Resume => 128,
            Self::Abort => 256,
            Self::Restart => 512,
            Self::Complete => 1024,
        }
    }
}

// ─── Operation / Source Mode ─────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum OperationMode {
    Offline,
    Operator,
    Automatic,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum SourceMode {
    Internal,
    External,
}

// ─── PEA Instance Runtime Status ─────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeaInstanceStatus {
    pub pea_id: String,
    pub deployed: bool,
    pub running: bool,
    pub services: Vec<ServiceRuntimeState>,
    pub opcua_endpoint: Option<String>,
    pub last_updated: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceRuntimeState {
    pub tag: String,
    pub state: ServiceState,
    pub current_procedure_id: Option<u32>,
    pub operation_mode: OperationMode,
    pub source_mode: SourceMode,
}

// ─── EVA-ICS Deployment Structures ───────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvaDeploymentPlan {
    pub pea_id: String,
    pub items: Vec<EvaItem>,
    pub controller_service: Option<EvaControllerConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvaItem {
    pub oid: String,
    pub enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvaControllerConfig {
    pub id: String,
    pub opcua_endpoint: String,
    pub node_mappings: Vec<OpcUaNodeMapping>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpcUaNodeMapping {
    pub oid: String,
    pub node_id: String,
    pub interval_ms: u32,
}

// ─── Recipe / Sequence Models ────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Recipe {
    pub id: String,
    pub name: String,
    pub description: String,
    pub steps: Vec<RecipeStep>,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecipeStep {
    pub order: u32,
    pub pea_id: String,
    pub service_tag: String,
    pub command: ServiceCommand,
    pub procedure_id: Option<u32>,
    pub parameters: Vec<RecipeParameterValue>,
    pub wait_for_state: Option<ServiceState>,
    pub timeout_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecipeParameterValue {
    pub parameter_tag: String,
    pub value: serde_json::Value,
}

// ─── Zenoh Topic Helpers ─────────────────────────────────────────────────────

pub mod topics {
    fn get_node_id() -> String {
        std::env::var("MURPH_NODE_ID").unwrap_or_else(|_| "local".to_string())
    }

    pub fn pea_announce(pea_id: &str) -> String {
        format!("murph/habitat/nodes/{}/pea/{}/announce", get_node_id(), pea_id)
    }

    pub fn pea_status(pea_id: &str) -> String {
        format!("murph/habitat/nodes/{}/pea/{}/status", get_node_id(), pea_id)
    }

    pub fn pea_service_state(pea_id: &str, service_tag: &str) -> String {
        format!(
            "murph/habitat/nodes/{}/pea/{}/services/{}/state",
            get_node_id(),
            pea_id,
            service_tag
        )
    }

    pub fn pea_service_command(pea_id: &str, service_tag: &str) -> String {
        format!(
            "murph/habitat/nodes/{}/pea/{}/services/{}/command",
            get_node_id(),
            pea_id,
            service_tag
        )
    }

    pub fn pea_data(pea_id: &str, data_tag: &str) -> String {
        format!(
            "murph/habitat/nodes/{}/pea/{}/data/{}",
            get_node_id(),
            pea_id,
            data_tag
        )
    }

    pub fn pea_config(pea_id: &str) -> String {
        format!("murph/habitat/nodes/{}/pea/{}/config", get_node_id(), pea_id)
    }

    pub fn pea_deploy(pea_id: &str) -> String {
        format!("murph/habitat/nodes/{}/pea/{}/deploy", get_node_id(), pea_id)
    }

    pub fn pea_lifecycle(pea_id: &str) -> String {
        format!("murph/habitat/nodes/{}/pea/{}/lifecycle", get_node_id(), pea_id)
    }

    pub const PEA_ANNOUNCE_WILDCARD: &str = "murph/habitat/nodes/*/pea/*/announce";
    pub const PEA_STATUS_WILDCARD: &str = "murph/habitat/nodes/*/pea/*/status";
    pub const PEA_DEPLOY_WILDCARD: &str = "murph/habitat/nodes/*/pea/*/deploy";
    pub const PEA_LIFECYCLE_WILDCARD: &str = "murph/habitat/nodes/*/pea/*/lifecycle";
    pub const PEA_SERVICE_COMMAND_WILDCARD: &str = "murph/habitat/nodes/*/pea/*/services/*/command";
    pub const POL_RECIPES_COMMAND: &str = "murph/pol/recipes/command";
    pub const POL_RECIPES_STATUS: &str = "murph/pol/recipes/status";
    pub const STATUS_EVA_ICS: &str = "murph/status/eva-ics";
}
